numba/function.py

```diff
@@ -0,0 +1,198 @@
+"""Provides Numba type, FunctionType, that makes functions as
+instances of a first-class function type.
+"""
+
+import types
+import numba
+from numba import types as nbtypes
+from numba.extending import typeof_impl
+from numba.extending import models, register_model
+from numba.extending import unbox, NativeValue, box
+from numba.targets.imputils import lower_constant
+from numba.ccallback import CFunc
+from numba import cgutils
+from llvmlite import ir
+from numba.types import (
+    FunctionType, FunctionPrototype, WrapperAddressProtocol)
+
+
+@typeof_impl.register(WrapperAddressProtocol)
+def typeof_WrapperAddressProtocol(val, c):
+    return val.signature().as_type()
+
+
+@typeof_impl.register(CFunc)
+def typeof_CFunc(val, c):
+    return val._sig.as_type()
+
+
+# TODO: typeof_impl for Dispatcher, types.FunctionType, ctypes.CFUNCTYPE
+
+
+@register_model(FunctionPrototype)
+class FunctionProtoModel(models.PrimitiveModel):
+    """FunctionProtoModel describes the signatures of first-class functions
+    """
+    def __init__(self, dmm, fe_type):
+        if isinstance(fe_type, FunctionType):
+            ftype = fe_type.ftype
+        elif isinstance(fe_type, FunctionPrototype):
+            ftype = fe_type
+        else:
+            raise NotImplementedError((type(fe_type)))
+        retty = dmm.lookup(ftype.rtype).get_value_type()
+        args = [dmm.lookup(t).get_value_type() for t in ftype.atypes]
+        be_type = ir.PointerType(ir.FunctionType(retty, args))
+        super(FunctionProtoModel, self).__init__(dmm, fe_type, be_type)
+
+
+@register_model(FunctionType)
+class FunctionModel(models.StructModel):
+    """FunctionModel holds addresses of function implementations
+    """
+    def __init__(self, dmm, fe_type):
+        members = [
+            # address of cfunc-ded function:
+            ('addr', nbtypes.voidptr),
+            # address of PyObject* referencing the Python function
+            # object, currently it is CFunc:
+            ('pyaddr', nbtypes.voidptr),
+        ]
+        super(FunctionModel, self).__init__(dmm, fe_type, members)
+
+
+@lower_constant(FunctionType)
+def lower_constant_function_type(context, builder, typ, pyval):
+    if isinstance(pyval, CFunc):
+        addr = pyval._wrapper_address
+        sfunc = cgutils.create_struct_proxy(typ)(context, builder)
+        llty = context.get_value_type(nbtypes.voidptr)
+        sfunc.addr = builder.inttoptr(ir.Constant(
```

Comment:
Can you switch the use of `builder.inttoptr` to `context.add_dynamic_addr`? Doing so will stop the compilation unit being cached to disk.

Reply:
This is only needed for freezing Constant pointers that are only valid in the current process.

Reply:
Done.